File: python/208_ImplementTrie.py
from unittest import TestCase
# https://leetcode.com/problems/implement-trie-prefix-tree/


class TrieNode:
    def __init__(self):
        self.end = False
        self.next = {}


class Trie:
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = TrieNode()

    def insert(self, word):
        """
        Inserts a word into the trie.
        :type word: str
        :rtype: void
        """
        node = self.root
        for ch in word:
            if not node.next.get(ch):
                node.next[ch] = TrieNode()
            node = node.next[ch]
        node.end = True

    def walk(self, word):
        node = self.root
        for ch in word:
            if not node.next.get(ch):
                return None
            node = node.next[ch]
        return node

    def search(self, word):
        """
        Returns if the word is in the trie.
        :type word: str
        :rtype: bool
        """
        node = self.walk(word)
        return bool(node and node.end)

    def startsWith(self, prefix):
        """
        Returns if there is any word in the trie that starts with the given prefix.
        :type prefix: str
        :rtype: bool
        """
        return bool(self.walk(prefix))


class Tests(TestCase):
    def test1(self):
        trie = Trie()
        trie.insert("apple")
        self.assertTrue(trie.search("apple"))
        self.assertTrue(not trie.search("app"))
        self.assertTrue(trie.startsWith("app"))
        trie.insert("app")
        self.assertTrue(trie.search("app"))

File: kotlin/src/main/kotlin/208.ImplementTrie.kt
import org.junit.Assert.assertEquals
import org.junit.Test
//https://leetcode.com/problems/implement-trie-prefix-tree/

class ImplementTrie{
  class Node{
    var end = false
    val next = HashMap<Char, Node>()
  }
  /** Initialize your data structure here. */
  val root = Node()

  /** Inserts a word into the trie. */
  fun insert(word: String) {
    var node = root
    for(ch in word){
      if(!node.next.containsKey(ch))
        node.next[ch] = Node()
      node = node.next[ch]!!
    }
    node.end = true
  }

  private fun walk(word: String): Node?{
    var node = root
    for(ch in word){
      if(node.next.containsKey(ch))
        node = node.next[ch]!!
      else
        return null
    }
    return node
  }

  /** Returns if the word is in the trie. */
  fun search(word: String): Boolean {
    val node = walk(word)
    return node != null && node.end
  }

  /** Returns if there is any word in the trie that starts with the given prefix. */
  fun startsWith(prefix: String) = walk(prefix) != null

  @Test
  fun test1(){
    insert("apple")
    assert(search("apple"))
    assert(!search("app"))
    assert(startsWith("app"))
    insert("app")
    assert(search("app"))
  }
}